services/solr-names-updater/src/solr_names_updater/worker.py

```diff
@@ -0,0 +1,143 @@
+
+The entry-point is the **cb_subscription_handler**
+
+The design and flow leverage a few constraints that are placed upon it
+by NATS Streaming and using AWAIT on the default loop.
+- NATS streaming queues require one message to be processed at a time.
+- AWAIT on the default loop effectively runs synchronously
+
+If these constraints change, the use of Flask-SQLAlchemy would need to change.
+Flask-SQLAlchemy currently allows the base model to be changed, or reworking
+the model to a standalone SQLAlchemy usage with an async engine would need
+to be pursued.
+"""
+import json
+import os
+
+import nats
+from flask import Flask
+from namex.models import db
+from queue_common.service import QueueServiceManager
+from queue_common.service_utils import QueueException, logger
+from requests import RequestException
+from sentry_sdk import capture_message
+from sqlalchemy.exc import OperationalError
+from urllib3.exceptions import NewConnectionError
+
+from solr_names_updater import config
+from solr_names_updater.names_processors.names import (
+    process_add_to_solr as process_names_add,
+    process_delete_from_solr as process_names_delete)
+from solr_names_updater.names_processors.possible_conflicts \
+    import (process_add_to_solr as process_possible_conflicts_add,
+            process_delete_from_solr as process_possible_conflicts_delete)
+
+
+qsm = QueueServiceManager()  # pylint: disable=invalid-name
+APP_CONFIG = config.get_named_config(os.getenv('DEPLOYMENT_ENV', 'production'))
+FLASK_APP = Flask(__name__)
+FLASK_APP.config.from_object(APP_CONFIG)
+db.init_app(FLASK_APP)
+
+
+def is_nr_state_change_msg_type(msg: dict):
+    """Check message is of type nr state change."""
+    if msg and msg.get('type', '') == 'bc.registry.nr_state.changed':
+        return True
+
+    return False
+
+
+def is_nr_furnished_change_msg_type(msg: dict):
+    """Check message is of type nr furnished changed."""
+    if msg and msg.get('type', '') == 'bc.registry.nr_furnished.changed':
+        return True
+
+    return False
+
+
+def is_processable(msg: dict):
+    """Determine if message is processable using message type of msg."""
+    if msg and \
+        (is_nr_state_change_msg_type(msg)
+         or is_nr_furnished_change_msg_type(msg)):
+        return True
+
+    return False
+
+
+async def process_nr_event_message(msg: dict, flask_app: Flask):
+    """Update solr accordingly based on incoming nr state changes."""
+    if not flask_app or not msg:
+        raise QueueException('Flask App or msg not available.')
+
+    with flask_app.app_context():
+        logger.debug('entering processing of nr event msg: %s', msg)
+
+        if is_nr_state_change_msg_type(msg):
+            data = msg.get('data')
+            new_state = data.get('newState')
+            if new_state in ('APPROVED', 'CONDITIONAL'):
+                process_names_add(data)
+            elif new_state in ('CANCELLED', 'CONSUMED'):
```

Comment:
do we need to implement something to handle a `RESET`?  and is this an actual NR state change or do we need to publish a different message with a different message type?

Reply:
yes if it is reset it needs to be removed from solr I believe

Reply:
is `RESET` an actual state?

Reply:
no it isn't so we will need to have a different way of checking for it

Reply:
it would probably help to see how the transaction history decides if the action was reset -- it isn't super well defined in their either so that might be something we need to look at

Reply:
A reset puts a name back into HOLD or INPROGRESS so it should get removed.

Reply:
@thorwolpert Reset isn’t actually a NR state. There is a *has_been_reset* field in requests table. Should we just use the NR state change message anyways and just treat it as state in this instance and publish a message by listening to the *requests.has_been_reset* field on the model?

Reply:
it does show as a state change from an approved state to hold/inprogress, so can be added tot he existing model listener

Reply:
Updated request model listener to publish `RESET` state change message when approved to hold/in progress state changes are detected.